v2/spanner-change-streams-to-sharded-file-sink/src/test/java/com/google/cloud/teleport/v2/templates/SpannerChangeStreamToGcsCustomShardIT.java

```diff
@@ -0,0 +1,259 @@
+import org.apache.beam.it.gcp.artifacts.Artifact;
+import org.apache.beam.it.gcp.spanner.SpannerResourceManager;
+import org.apache.beam.it.gcp.spanner.conditions.SpannerRowsCheck;
+import org.apache.beam.it.gcp.storage.GcsResourceManager;
+import org.apache.beam.it.gcp.storage.conditions.GCSArtifactsCheck;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Integration test for multiple shard database with custom shard id. */
+@Category({TemplateIntegrationTest.class, SkipDirectRunnerTest.class})
+@TemplateIntegrationTest(SpannerChangeStreamsToShardedFileSink.class)
+@RunWith(JUnit4.class)
+public class SpannerChangeStreamToGcsCustomShardIT extends SpannerChangeStreamToGcsITBase {
+  private static final Logger LOG = LoggerFactory.getLogger(SpannerChangeStreamToGcsSimpleIT.class);
+  private static HashSet<SpannerChangeStreamToGcsCustomShardIT> testInstances = new HashSet<>();
+
+  private static GcsResourceManager gcsResourceManager;
+  private static SpannerResourceManager spannerResourceManager;
+  private static SpannerResourceManager spannerMetadataResourceManager;
+  private static final String spannerDdl =
+      "SpannerChangeStreamToGcsCustomShardIT/spanner-schema.sql";
+  private static final String sessionFileResourceName =
+      "SpannerChangeStreamToGcsCustomShardIT/session.json";
+  private static PipelineLauncher.LaunchInfo jobInfo;
+
+  /**
+   * Does the following setup:
+   *
+   * <p>1. Creates a Spanner database with a given table 2. Creates a shard file with the connection
+   * details 3. The session file for the same is taken from the resources and uploaded to GCS 4.
+   * Places the session file and shard file in GCS 5. Creates the change stream in Spanner database
+   * 6. Creates the metadata database 7. Creates a JAR of v2/spanner-custom-shard and uploads it to
+   * GCS 8. Launches the job to read from Spanner and write to GCS
+   *
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  @Before
+  public void setUp() throws IOException, InterruptedException {
+    skipBaseCleanup = true;
+    synchronized (SpannerChangeStreamToGcsSimpleIT.class) {
+      testInstances.add(this);
+      if (jobInfo == null) {
+        gcsResourceManager = createGcsResourceManager(getClass().getSimpleName());
+        spannerResourceManager = createSpannerResourceManager();
+        createSpannerDatabase(spannerResourceManager, spannerDdl);
+        uploadSessionFileToGcs(gcsResourceManager, sessionFileResourceName);
+        spannerMetadataResourceManager = createSpannerMetadataResourceManager();
+        createSpannerMetadataDatabase(spannerMetadataResourceManager);
+        createAndUploadJarToGcs(gcsResourceManager);
+        createAndUploadShardConfigToGcs();
+        launchReaderDataflowJob();
```

Comment:
Can we further encapsulate this into a parameterized `launch` or `prelaunch` method? Looks like a lot of method calls for things that will essentially always need to be done. 

@darshan-sj You did something similar for live migration, do you have a suggestion here?

Reply:
Looks good. Currrently all the methods in Base class are like utility methods. If we have to compress this further, we will have to move the fields gcsResourceManager, and other resource managers to base class. I would prefer this implementation over that.